### ISSUE TITLE
Possibility to see which tabs have new output?

### ISSUE THREAD

**Issue by kramer65** [gnunn1/tilix#1445]
In iTerm2 on MacOS it is really easy to see which tabs have new output. In the screenshot below you see that the tabs with a blue dot (the first and third tab) have new output. Is it possible to have something similar in tilix?

![screen shot 2017-03-04 at 11 59 35 am](https://user-images.githubusercontent.com/596581/41718366-0634b5da-755d-11e8-9345-02c2ed4924e2.png)

**Comment by kramer65** [comment]
When I was still on MacOS I used this a lot for force tailing logs in multiple tabs. Using these blue dots easily alerted me to the tab which had some new info. I guess the notification on ending a command would not work because `tail -f` never ends.

So it would be really really cool if this could be added! :-D

**Comment by gnunn1** [comment]
I have added this feature, however the VTE has a limitation in that it triggers the signals you could use to determine new output was inserted (onContentsChanged or onTextInserted) from things other then new output such as resizing the window. This means that in order to track whether new text was inserted tilix has to track the last row,col position.

This entails the same problem that triggers have in tilix, which is that alternate screen mode uses a different set of coordinates for row,col which means that tilix needs to be aware of when alternate screen mode is in use. The OOTB VTE does not provide a mechanism for doing this, I have a patch for VTE which adds a signal for it.

So long story short, this feature requires a patched VTE. On Arch Linux this is available in the vte3-tilix package but it is the only distribution that includes it.

@egmontkob if you have some thoughts on this let me know, I may be over-complicating the solution if I am missing something.

Here is a screenshot of the feature in action:

![screenshot from 2018-06-23 18-48-33](https://user-images.githubusercontent.com/9358238/41814343-380528ae-7717-11e8-9858-887315550f07.png)

**Comment by egmontkob** [comment]
I'm not aware of the technical details, but I guess watching for onContentsChanged and onTextInserted should be enough. What else is missing? Are these not emitted e.g. when changing between screens? Then this should be fixed in VTE.

I don't understand how window resizing is relevant. First, most of the time it's a user-initiated event, hence Tilix knows it anyway even without a signal. In some cases it's also possible to initiate it via an escape sequence, I don't know if Tilix supports it, if it does then it must be done via connecting to a signal, so there you go, you have the signal. Anyway, if no additional text appears on the screen (only the previous contents are reflown) upon a resize then you probably don't want to notify. And if the screen, or part thereof is redrawn by an app then you get ContentsChanged or TextInserted. What am I missing?

Similarly I don't see how tracking the cursor position is necessary/important. But again I haven't checked the details and I might easily be missing something.

**Comment by egmontkob** [comment]
[711059](https://bugzilla.gnome.org/show_bug.cgi?id=711059) is the main gnome-terminal bug tracking notification support. It's long and inconclusive. The only probably clear consensus is that there are two significantly different ways of doing notifications.

One is the "implicit", activity based one, mostly for tab coloring. I guess you should be able to do it already with unpatched VTE, and this is what this bug seems to ask for.

Another one is the "explicit" via an escape sequence or similar. This is where Fedora went ahead and did _something_, which is IMO heavily buggy (and those bugs can't be fixed without buy-in from bash, whose maintainer wasn't supportive), hence my resistance in applying it mainstream.

**Comment by gnunn1** [comment]
The problem is that neither onTextInserted or onContentsChanged are fired solely based on new text being inserted, both of them get triggered by other things unrelated to new text. For example, resizing the window, and thus VTE, causes both events to be fired. This makes it impossible to tell if the event is being triggered because there is new text inserted or because of something else like resizing.

For onContentsChanged it makes sense since the contents will change when resizing due to reflow and other things. onTextInserted though is more surprising, however looking at the docs this one looks to be more targeted at accessibility then the use case I'm looking at.

That notification issue isn't particularly relevant here, the use case here is I'm monitoring a process (i.e. tailing a log) while working on a terminal in a different session. I want a visual indicator of when new output is written in the terminal for sessions that are not in view. Notifications are for when the command completes, the notification will never fire when monitoring a log file since the command prompt never comes back.

I can open a bug in VTE if you think it's worthwhile, just checking that I'm not missing anything.

**Comment by egmontkob** [comment]
> For example, resizing the window, and thus VTE, causes both events to be fired.

I see. I don't think it should be this way. I agree the reflow triggered by the resize shouldn't generate TextInserted.

[If an app subsequently (due to the resize) decides to reprint some characters then of course TextInserted is inevitable. But that's not your concern.]

> I can open a bug in VTE

Yes please :)
